File: app/src/main/java/com/kickstarter/services/apirequests/RegisterWithFacebookBody.java
package com.kickstarter.services.apirequests;

import android.os.Parcelable;

import com.kickstarter.libs.qualifiers.AutoGson;

import auto.parcel.AutoParcel;

@AutoGson
@AutoParcel
public abstract class RegisterWithFacebookBody implements Parcelable {
  public abstract String accessToken();
  public abstract boolean sendNewsletters();

  @AutoParcel.Builder
  public abstract static class Builder {
    public abstract Builder accessToken(String __);
    public abstract Builder sendNewsletters(boolean __);
    public abstract RegisterWithFacebookBody build();
  }

  public static Builder builder() {
    return new AutoParcel_RegisterWithFacebookBody.Builder();
  }

  public abstract Builder toBuilder();

}

File: app/src/main/java/com/kickstarter/services/apirequests/RegisterWithFacebookBody.kt
package com.kickstarter.services.apirequests

import android.os.Parcelable
import kotlinx.parcelize.Parcelize

@Parcelize
class RegisterWithFacebookBody private constructor(
    private val accessToken: String,
    private val newsletterOptIn: Boolean,
    private val sendNewsletters: Boolean,
) : Parcelable {
    fun accessToken() = this.accessToken
    fun newsletterOptIn() = this.newsletterOptIn
    fun sendNewsletters() = this.sendNewsletters

    @Parcelize
    data class Builder(
        private var accessToken: String = "",
        private var newsletterOptIn: Boolean = false,
        private var sendNewsletters: Boolean = false
    ) : Parcelable {
        fun accessToken(accessToken: String) = apply { this.accessToken = accessToken }
        fun newsletterOptIn(newsletterOptIn: Boolean) = apply { this.newsletterOptIn = newsletterOptIn }
        fun sendNewsletters(sendNewsletters: Boolean) = apply { this.sendNewsletters = sendNewsletters }
        fun build() = RegisterWithFacebookBody(
            accessToken = accessToken,
            newsletterOptIn = newsletterOptIn,
            sendNewsletters = sendNewsletters
        )
    }

    override fun equals(obj: Any?): Boolean {
        var equals = super.equals(obj)
        if (obj is RegisterWithFacebookBody) {
            equals = accessToken() == obj.accessToken() &&
                newsletterOptIn() == obj.newsletterOptIn() &&
                sendNewsletters() == obj.sendNewsletters()
        }
        return equals
    }

    fun toBuilder() = Builder(
        accessToken = accessToken,
        newsletterOptIn = newsletterOptIn,
        sendNewsletters = sendNewsletters
    )

    companion object {
        @JvmStatic
        fun builder() = Builder()
    }
}
